### PR TITLE
Refresh System.Runtime ref from System.Private.CoreLib in Coreclr

### PR DESCRIPTION
**Pull request by weshaggard**
This also includes an update to the genapi tool that includes the
necessary private fields in structs to enable the compiler to
do correctness checks.

See https://github.com/dotnet/corefx/issues/6185.

The ref was generated with the updated changes in https://github.com/dotnet/buildtools/pull/1859. 

PTAL @jkotas @VSadov @jaredpar @gafter 

Note that some of the updates are just a refresh from what is in the System.Private.CoreLib and not about the fields in structs.

### CODE REVIEW

File: src/System.Runtime/ref/System.Runtime.cs [hunk 1]
```diff
@@ -7897,6 +7964,7 @@ public partial class UnobservedTaskExceptionEventArgs : System.EventArgs
     [System.Runtime.CompilerServices.AsyncMethodBuilderAttribute(typeof(System.Runtime.CompilerServices.AsyncValueTaskMethodBuilder<>))]
     public readonly partial struct ValueTask<TResult> : System.IEquatable<System.Threading.Tasks.ValueTask<TResult>>
     {
+        internal readonly TResult _result;
```
Comment (jkotas) [inline src/System.Runtime/ref/System.Runtime.cs]: Should all the fields be private?

File: src/System.Runtime/ref/System.Runtime.cs [hunk 2]
```diff
@@ -4991,11 +5044,11 @@ public partial class FileStream : System.IO.Stream
         public virtual System.IntPtr Handle { get { throw null; } }
         public virtual bool IsAsync { get { throw null; } }
         public override long Length { get { throw null; } }
-        public string Name { get { throw null; } }
+        public virtual string Name { get { throw null; } }
```
Comment (jkotas) [inline src/System.Runtime/ref/System.Runtime.cs]: There is a bunch of unrelated changes like this one. Are they all intentional?

File: src/System.Runtime/ref/System.Runtime.cs [hunk 3]
```diff
@@ -1952,11 +1968,11 @@ public partial class RankException : System.SystemException
         [System.ComponentModel.EditorBrowsableAttribute((System.ComponentModel.EditorBrowsableState)(1))]
         public static System.ReadOnlySpan<T> DangerousCreate(object obj, ref T objectData, int length) { throw null; }
         [System.ComponentModel.EditorBrowsableAttribute((System.ComponentModel.EditorBrowsableState)(1))]
-        [System.ObsoleteAttribute("Equals() on ReadOnlySpan will always throw an exception. Use == instead.")]
+        [System.ObsoleteAttribute("Equals() on Span will always throw an exception. Use == instead.")]
```
Comment (jkotas) [inline src/System.Runtime/ref/System.Runtime.cs]: The original message looks right. This looks like a bug in the implementation.

Comment (weshaggard) [reply src/System.Runtime/ref/System.Runtime.cs]: This just updated to what is in the implementation. We could definitely update both to be consistent with ReadOnlySpan.

File: src/System.Runtime/ref/System.Runtime.cs (hunk 2 above)
Comment (weshaggard) [reply src/System.Runtime/ref/System.Runtime.cs]: Yes these are intentional. They are essentially refreshing based on what is in System.Private.CoreLib implementation. A number of changes to the API never made it to the ref.

File: src/System.Runtime/ref/System.Runtime.cs (hunk 1 above)
Comment (weshaggard) [reply src/System.Runtime/ref/System.Runtime.cs]: I noticed a couple of these as well, I don't know why they showed up as internal but I will investigate.

File: src/System.Runtime/ref/System.Runtime.cs [hunk 4]
```diff
@@ -6769,6 +6835,7 @@ public sealed partial class FieldOffsetAttribute : System.Attribute
     }
     public partial struct GCHandle
     {
+        private object _dummy;
```
Comment (jkotas) [inline src/System.Runtime/ref/System.Runtime.cs]: I think this should be `int`. GCHandle does not have instance object fields.

File: src/System.Runtime/ref/System.Runtime.cs (hunk 3 above)
Comment (jkotas) [reply src/System.Runtime/ref/System.Runtime.cs]: Fix: https://github.com/dotnet/coreclr/pull/15820

File: src/System.Runtime/ref/System.Runtime.cs (hunk 4 above)
Comment (weshaggard) [reply src/System.Runtime/ref/System.Runtime.cs]: this might be a side-effect of me not filtering out statics while doing the contains reference check. I will regenerate once I update the tool.

Comment (weshaggard) [reply src/System.Runtime/ref/System.Runtime.cs]: @jkotas the GCHandle holds a IntPtr which in turn holds a void* so based on our logic I would expect this to have an object dummy field. Is there something I'm missing?

File: src/System.Runtime/ref/System.Runtime.cs (hunk 1 above)
Comment (weshaggard) [reply src/System.Runtime/ref/System.Runtime.cs]: The reason this and a couple others are internal instead of private is because for any non-public fields that are generic I just blindly include those unchanged. For now I will keep them like this so the correctly show the usage of generic parameters. If you feel there is a reason to switch them I can.

Comment (jkotas) [reply src/System.Runtime/ref/System.Runtime.cs]: ok

File: src/System.Runtime/ref/System.Runtime.cs (hunk 4 above)
Comment (jkotas) [reply src/System.Runtime/ref/System.Runtime.cs]: `void*` should be treated as `int`, not as `object`.

Comment (weshaggard) [reply src/System.Runtime/ref/System.Runtime.cs]: OK with that change we get a bunch of updates. Can you have a look over https://github.com/dotnet/corefx/pull/26264/commits/9288c499b2491a2208225b8621ac99f7a301081e to ensure they look as expected?

File: src/System.Runtime/ref/System.Runtime.cs [hunk 5]
```diff
@@ -404,6 +405,7 @@ public partial class BadImageFormatException : System.SystemException
     }
     public partial struct Boolean : System.IComparable, System.IComparable<bool>, System.IConvertible, System.IEquatable<bool>
     {
+        private object _dummy;
```
Comment (VSadov) [inline src/System.Runtime/ref/System.Runtime.cs]: Why Boolean contains an object field?
The real one only contains a bool.

File: src/System.Runtime/ref/System.Runtime.cs [hunk 6]
```diff
@@ -488,6 +491,7 @@ public partial struct Byte : System.IComparable, System.IComparable<byte>, Syste
     }
     public partial struct Char : System.IComparable, System.IComparable<char>, System.IConvertible, System.IEquatable<char>
     {
+        private object _dummy;
```
Comment (VSadov) [inline src/System.Runtime/ref/System.Runtime.cs]: Same as Boolean - why object?

File: src/System.Runtime/ref/System.Runtime.cs (hunk 5 above)
Comment (weshaggard) [reply src/System.Runtime/ref/System.Runtime.cs]: Just pushed a change that fixes this.

File: src/System.Runtime/ref/System.Runtime.cs (hunk 6 above)
Comment (weshaggard) [reply src/System.Runtime/ref/System.Runtime.cs]: Was an issue in detecting a reference type, just fixed.

File: src/System.Runtime/ref/System.Runtime.cs [hunk 7]
```diff
@@ -1557,6 +1567,7 @@ public partial struct Int64 : System.IComparable, System.IComparable<long>, Syst
     }
     public partial struct IntPtr : System.IEquatable<System.IntPtr>, System.Runtime.Serialization.ISerializable
     {
+        private object _dummy;
```
Comment (VSadov) [inline src/System.Runtime/ref/System.Runtime.cs]: Should be `int` or IntPtr.

File: src/System.Runtime/ref/System.Runtime.cs [hunk 8]
```diff
@@ -3204,10 +3256,10 @@ public partial struct UInt64 : System.IComparable, System.IComparable<ulong>, Sy
     [System.CLSCompliantAttribute(false)]
     public partial struct UIntPtr : System.IEquatable<System.UIntPtr>, System.Runtime.Serialization.ISerializable
     {
+        private object _dummy;
```
Comment (VSadov) [inline src/System.Runtime/ref/System.Runtime.cs]: should be `int` or UIntPtr

File: src/System.Runtime/ref/System.Runtime.cs [hunk 9]
```diff
@@ -2005,10 +2025,12 @@ public partial struct RuntimeMethodHandle : System.Runtime.Serialization.ISerial
     }
     public partial struct RuntimeTypeHandle : System.Runtime.Serialization.ISerializable
     {
+        private object _dummy;
         public System.IntPtr Value { get { throw null; } }
         public override bool Equals(object obj) { throw null; }
         public bool Equals(System.RuntimeTypeHandle handle) { throw null; }
         public override int GetHashCode() { throw null; }
+        [System.CLSCompliantAttribute(false)]
```
Comment (jkotas) [inline src/System.Runtime/ref/System.Runtime.cs]: You can delete this unnecessary CLSCompliantAttribute attribute. I have deleted it in the implementation already.

File: src/System.Runtime/ref/System.Runtime.cs [hunk 10]
```diff
@@ -7498,6 +7563,7 @@ namespace System.Threading
 {
     public readonly partial struct CancellationToken
     {
+        private readonly object _dummy;
```
Comment (VSadov) [inline src/System.Runtime/ref/System.Runtime.cs]: 👍 This one was causing particularly annoying bug reports.

**Comment by VSadov** [comment]
>         public static System.Span<T> DangerousCreate(object obj, ref T objectData, int length) { throw null; }

Just wondering - what happened to the `DangerousGetPinnableReference` ?

---
Refers to: src/System.Runtime/ref/System.Runtime.cs:2170 in 925866d. [](commit_id = 925866d664d3ff2ef49b12243e117413782ae0ae, deletion_comment = False)

File: src/System.Runtime/ref/System.Runtime.cs (hunk 1 above)
Comment (VSadov) [reply src/System.Runtime/ref/System.Runtime.cs]: Same in `Nullable<T>`    -   `internal T value` .   I think that is ok. 

---
In reply to: [161028771](https://github.com/dotnet/corefx/pull/26264#discussion_r161028771) [](ancestors = 161028771)

**Comment by weshaggard** [comment]
>Just wondering - what happened to the DangerousGetPinnableReference ?

https://github.com/dotnet/coreclr/commit/4d0e3632cdf940a2170e2cf7469d6bf40483c90b

File: src/System.Runtime/ref/System.Runtime.cs [hunk 11]
```diff
@@ -2131,6 +2155,7 @@ public partial struct Single : System.IComparable, System.IComparable<float>, Sy
     }
     public readonly ref partial struct Span<T>
     {
+        private readonly int _dummy;
```
Comment (VSadov) [inline src/System.Runtime/ref/System.Runtime.cs]: I think .NET implementation contains an object. And even the fast implementation technically may contain an object reference.

We do not consider managed structs as unmanaged right now, but might in the future. 

I think Span/ReadOnlySpan should be considered object-containing

Comment (jkotas) [reply src/System.Runtime/ref/System.Runtime.cs]: > I think Span/ReadOnlySpan should be considered object-containing

I agree. It is not something that the tool can figure out from CoreLib metadata because of Span is treated as intrinsic by the runtime. Types like these will need exceptions. 

BTW: It is a bit odd that the primitive types like Int64 has 32-bit fields now. I won't be surprised if we find that it broke something somewhere and we need similar exceptions for them.

Comment (VSadov) [reply src/System.Runtime/ref/System.Runtime.cs]: On the other hand - even if we see all the fields we would consider fast span unmanaged. (since we do not see or understand the hidden byref)
And other ones - like `TypedReference` were always considered unmanaged.

I guess the object inside the slow spans would be the only reason to make the spans managed. - Just to conservatively match  the implementations.


---
In reply to: [161045136](https://github.com/dotnet/corefx/pull/26264#discussion_r161045136) [](ancestors = 161045136)

Comment (VSadov) [reply src/System.Runtime/ref/System.Runtime.cs]: Yeah - I had that concern about primitives as well. Maybe it would be safer if they are self-containing, like all their implementations?

---
In reply to: [161048209](https://github.com/dotnet/corefx/pull/26264#discussion_r161048209) [](ancestors = 161048209)

Comment (VSadov) [reply src/System.Runtime/ref/System.Runtime.cs]: I.E.  -   `int`contains an `int` dummy, `bool` contains a `bool`, and so on ...

---
In reply to: [161049185](https://github.com/dotnet/corefx/pull/26264#discussion_r161049185) [](ancestors = 161049185,161048209)

**Comment by jkotas** [comment]
> Just wondering - what happened to the `DangerousGetPinnableReference` ?

It was replaced by `System.Runtime.InteropServices.MemoryMarshal.GetReference`, and removed from the public surface. The DangerousCreate method will be moved to MemoryMarshal as well before we ship - there is an issue about that.

Once the pattern-based fixed statement in C# is figured out, we will add it back in the right form:
https://github.com/dotnet/csharplang/pull/1100#issuecomment-353763284

File: src/System.Runtime/ref/System.Runtime.cs (hunk 11 above)
Comment (jkotas) [reply src/System.Runtime/ref/System.Runtime.cs]: > And other ones - like TypedReference were always considered unmanaged.

While it actually contains managed references because of it is treated as intrinsic by the runtime...

This highlights that this has number of false positives and false negatives.

Would it make sense to relax C# to let you take pointer from anything, and demote the current error for taking pointer of "managed type" to just a warning?

**Comment by VSadov** [comment]
Re: `DangerousGetPinnableReference` - makes sense. Thanks!

File: src/System.Runtime/ref/System.Runtime.cs (hunk 11 above)
Comment (VSadov) [reply src/System.Runtime/ref/System.Runtime.cs]: Possibly. I do not know the exact rationale for not allowing pointers to managed structs. It was the case since v1.0

---
In reply to: [161052501](https://github.com/dotnet/corefx/pull/26264#discussion_r161052501) [](ancestors = 161052501)

Comment (weshaggard) [reply src/System.Runtime/ref/System.Runtime.cs]: Latest update fixes primitive types to have matching field types and manually marks Span and ReadOnlySpan as having an object type.

Comment (weshaggard) [reply src/System.Runtime/ref/System.Runtime.cs]: Are there any other special types we need to manually fix?

Comment (VSadov) [reply src/System.Runtime/ref/System.Runtime.cs]: I think this is all that is needed. 
The old-style ref-likes  like TypedReference should stay as they are. At least for compat reasons.

---
In reply to: [161073853](https://github.com/dotnet/corefx/pull/26264#discussion_r161073853) [](ancestors = 161073853)

File: src/System.Runtime/ref/System.Runtime.cs [hunk 12]
```diff
@@ -1968,6 +1984,7 @@ public partial class RankException : System.SystemException
         public bool TryCopyTo(System.Span<T> destination) { throw null; }
         public ref partial struct Enumerator
         {
+            private int _dummy;
```
Comment (jkotas) [inline src/System.Runtime/ref/System.Runtime.cs]: The Enumerator struct contains Span field. It should be treated same way as Span and have object field.

This applies to all structs with Span fields.

File: src/System.Runtime/ref/System.Runtime.cs [hunk 13]
```diff
@@ -2162,6 +2186,7 @@ public partial struct Single : System.IComparable, System.IComparable<float>, Sy
...
```
Comment (jkotas) [inline src/System.Runtime/ref/System.Runtime.cs]: Same here

**Comment by weshaggard** [comment]
test Linux x64 Release Build
test UWP CoreCLR x64 Debug Build

File: src/System.Runtime/ref/System.Runtime.cs (hunk 12 above)
Comment (weshaggard) [reply src/System.Runtime/ref/System.Runtime.cs]: I should probably update the tool to special case Span and ReadOnlySpan.

Comment (VSadov) [reply src/System.Runtime/ref/System.Runtime.cs]: Perhaps it is easier to hack the tool so that it thinks Span/ReadonlySpan actually contain an extra `private object _dummy` ?

File: src/System.Runtime/ref/System.Runtime.cs [hunk 14]
```diff
@@ -1940,6 +1955,7 @@ public partial class RankException : System.SystemException
     }
     public readonly ref partial struct ReadOnlySpan<T>
     {
+        private readonly object _dummy; //Manually set to object because tools don't know it is special
```
Comment (weshaggard) [inline src/System.Runtime/ref/System.Runtime.cs]: @jkotas @VSadov should this be of type object or type T?

Comment (VSadov) [reply src/System.Runtime/ref/System.Runtime.cs]: `object` , there are no fields of type `T` in Span.

Comment (VSadov) [reply src/System.Runtime/ref/System.Runtime.cs]: It is similar to the `ImmutableArray<T>` - `T` is the element type, but the struct is just an indirection to some other storage, it does not contain any instances directly.

File: src/System.Runtime/ref/System.Runtime.cs (hunk 12 above)
Comment (weshaggard) [reply src/System.Runtime/ref/System.Runtime.cs]: I'm going to teach the tool to recognize "ByReference<T>" as hold a reference.

Comment (VSadov) [reply src/System.Runtime/ref/System.Runtime.cs]: That could add an object into `TypedReference`. 
It is technically breaking. `TypedReference*` It is a very corner case and maybe fragile/broken anyways, so I am not sure how much we care.

Comment (weshaggard) [reply src/System.Runtime/ref/System.Runtime.cs]: The rule https://github.com/dotnet/buildtools/pull/1859/commits/4b80362df58e1ef45e58ca482b718084baddbcf2 produced the expected output 
https://github.com/dotnet/corefx/pull/26264/commits/6ef954562c71fe7f3de4db7407d3d4ec58825dd7.

TypedReference didn't change because it only contains an IntPtr not a ByReference.

File: src/System.Runtime/ref/System.Runtime.cs (hunk 14 above)
Comment (weshaggard) [reply src/System.Runtime/ref/System.Runtime.cs]: I have the changed for ImmutableArray locally and my tool generates `internal T[] array;` for it as it is generic underneath.

Comment (VSadov) [reply src/System.Runtime/ref/System.Runtime.cs]: Either `T[]` or `object` would work since `T[]` is not a struct.

Comment (VSadov) [reply src/System.Runtime/ref/System.Runtime.cs]: Basically arrays are just reference types for this purpose.
